web3.js/src/connection.ts

```diff
@@ -4263,75 +4207,273 @@ export class Connection {
+    serverSubscriptionId: ServerSubscriptionId,
+    publishArgs: Parameters<TCallback>,
+  ): void {
+    const callbacks =
+      this._subscriptionCallbacksByServerSubscriptionId[serverSubscriptionId];
+    if (callbacks == null) {
+      return;
     }
-
-    for (let clientSubscriptionId of logsClientSubscriptionIds) {
-      const sub = this._logsSubscriptions[clientSubscriptionId];
-      let filter;
-      if (typeof sub.filter === 'object') {
-        filter = {mentions: [sub.filter.toString()]};
-      } else {
-        filter = sub.filter;
+    callbacks.forEach(cb => {
+      try {
+        cb(...(publishArgs as [any, any]));
+      } catch (e) {
+        console.error(e);
       }
-      this._subscribe(
-        sub,
-        'logsSubscribe',
-        this._buildArgs([filter], sub.commitment),
-      );
-    }
+    });
   }
 
   /**
    * @internal
    */
   _wsOnAccountNotification(notification: object) {
-    const res = create(notification, AccountNotificationResult);
-    for (const sub of Object.values(this._accountChangeSubscriptions)) {
-      if (sub.subscriptionId === res.subscription) {
-        sub.callback(res.result.value, res.result.context);
+    const {result, subscription} = create(
+      notification,
+      AccountNotificationResult,
+    );
+    this._publishNotification<AccountChangeCallback>(subscription, [
+      result.value,
+      result.context,
+    ]);
+  }
+
+  /**
+   * @private
+   */
+  _makeSubscription(
+    // The subscription config, before applying defaults
+    rawConfig: SubscriptionConfig,
+  ): ClientSubscriptionId {
+    const clientSubscriptionId = this._nextClientSubscriptionId++;
+    let subscriptionConfig: SubscriptionConfig;
+    // Apply defaults.
+    switch (rawConfig.method) {
+      case 'accountSubscribe':
+      case 'logsSubscribe':
+      case 'programSubscribe':
+        if (rawConfig.params.commitment === undefined) {
+          subscriptionConfig = {
+            ...rawConfig,
+            params: {
+              ...rawConfig.params,
+              commitment: 'finalized',
+            },
+          } as SubscriptionConfig;
+        } else {
+          subscriptionConfig = rawConfig;
+        }
+        break;
+      case 'signatureSubscribe':
+        if (
+          rawConfig.params.options &&
+          rawConfig.params.options.commitment === undefined
+        ) {
+          subscriptionConfig = {
+            ...rawConfig,
+            params: {
+              ...rawConfig.params,
+              options: {
+                ...rawConfig.params.options,
+                commitment: 'finalized',
+              },
+            },
+          };
+        } else {
+          subscriptionConfig = rawConfig;
+        }
+        break;
+      default:
+        subscriptionConfig = rawConfig;
+    }
+    const hash = hashSubscriptionConfig(subscriptionConfig);
+    const existingSubscription = this._subscriptionsByHash[hash];
+    if (existingSubscription == null) {
```

Comment:
Yeah, this is one of those funny things. It's quite common among folks to check for `null | undefined` using the coercive `== null`. Here, you have a type that TypeScript damn well knows to be `Subscription | undefined`, but despite that it doesn't complain about the use of a coercive `null` check, because a coercive `null` check is valid here.

I'll change it, though.